File: Difficulty: Easy/Parenthesis Checker/parenthesis-checker.java
//{ Driver Code Starts
import java.util.*;
import java.io.*;
import java.lang.*;

class Driverclass
{
    public static void main(String args[])
    {
        Scanner sc = new Scanner(System.in);
        
        //Reading total number of testcases
        int t= sc.nextInt();
        
        while(t-- >0)
        {
            //reading the string
            String st = sc.next();
            
            //calling ispar method of Paranthesis class 
            //and printing "balanced" if it returns true
            //else printing "not balanced"
            if(new Solution().ispar(st) == true)
                System.out.println("balanced");
            else
                System.out.println("not balanced");
        
        }
    }
}
// } Driver Code Ends



class Solution
{
    //Function to check if brackets are balanced or not.
    public boolean ispar(String s) {
        if(s.length()<=1){
            return false;
        }
        Stack<Character> stack = new Stack<>();
        for(int i=0;i<s.length();i++){
            if(s.charAt(i)=='[' || s.charAt(i)=='{' || s.charAt(i)=='('){
                stack.push(s.charAt(i));
            }
            else if(stack.empty()){
                return false;
            }
            else{
                if(s.charAt(i)==')'){
                    if(stack.pop()!='('){
                        return false;
                    }
                }
                if(s.charAt(i)=='}'){
                    if(stack.pop()!='{'){
                        return false;
                    }
                }
                if(s.charAt(i)==']'){
                    if(stack.pop()!='['){
                        return false;
                    }
                }
            }
        }
        if (stack.empty()){
            return true;
        }
        return false;
    }
}

File: Difficulty: Easy/Parenthesis Checker/parenthesis-checker.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution {
  public:
    bool isBalanced(string& s) {
        // code here
        int n = s.size();
        stack<char> ss;
        for (int i = 0; i < n; i++) {
            char ch = s[i];
            if (ch == '(' || ch == '{' || ch == '[') {
                ss.push(s[i]);
                continue;
            } 
            else if (ss.empty() || (ch == ')' && ss.top() != '(') || (ch == ']' && ss.top() != '[') || (ch == '}' && ss.top() != '{')) {
                return false;
            } 
            else {
                ss.pop();
            }
        }
        return ss.empty();
    }
};

//{ Driver Code Starts.

int main() {
    int t;
    string a;
    cin >> t;
    while (t--) {
        cin >> a;
        Solution obj;
        if (obj.isBalanced(a))
            cout << "true" << endl;
        else
            cout << "false" << endl;

        cout << "~"
             << "\n";
    }
}
// } Driver Code Ends